tools/rust-tools/run-manager/src/main.rs

```diff
@@ -0,0 +1,393 @@
+            let stderr = String::from_utf8_lossy(&stop_output.stderr);
+            error!("Warning: Docker stop failed: {}", stderr);
+        }
+
+        // Remove the container
+        let rm_output = Command::new("docker")
+            .arg("rm")
+            .arg(container_id)
+            .output()
+            .context("Failed to remove container")?;
+
+        if !rm_output.status.success() {
+            let stderr = String::from_utf8_lossy(&rm_output.stderr);
+            error!("Warning: Docker rm failed: {}", stderr);
+        }
+
+        Ok(())
+    }
+}
+
+/// Coordinator client for querying Solana
+struct CoordinatorClient {
+    rpc_client: RpcClient,
+    #[allow(dead_code)]
+    program_id: Pubkey,
+}
+
+impl CoordinatorClient {
+    fn new(rpc_endpoint: String, program_id: Pubkey) -> Self {
+        let rpc_client = RpcClient::new(rpc_endpoint);
+        Self {
+            rpc_client,
+            program_id,
+        }
+    }
+
+    // Fetch coordinator data and deserialize into a struct
+    fn fetch_coordinator_data(&self, run_id: &str) -> Result<CoordinatorInstance> {
+        // Derive the coordinator instance PDA
+        let coordinator_instance = find_coordinator_instance(run_id);
+
+        let account = self
+            .rpc_client
+            .get_account(&coordinator_instance)
+            .context("RPC error: failed to get account")?;
+
+        let instance = CoordinatorInstance::try_deserialize(&mut account.data.as_slice())
+            .context("Failed to deserialize CoordinatorInstance")?;
+
+        Ok(instance)
+    }
+
+    fn get_docker_tag_for_run(&self, run_id: &str) -> Result<String> {
+        info!("Querying coordinator for Run ID: {}", run_id);
+
+        let instance = self.fetch_coordinator_data(run_id)?;
+
+        // Fetch the coordinator account to get the client version
+        let coordinator_account_data = self
+            .rpc_client
+            .get_account(&instance.coordinator_account)
+            .context("RPC error: failed to get coordinator account")?;
+
+        let coordinator_account = coordinator_account_from_bytes(&coordinator_account_data.data)
+            .context("Failed to deserialize CoordinatorAccount")?;
+
+        let client_version = String::from(&coordinator_account.state.client_version);
+
+        info!(
+            "Fetched CoordinatorInstance from chain: {{ run_id: {}, coordinator_account: {}, client_version: {} }}",
+            instance.run_id, instance.coordinator_account, client_version
+        );
+
+        let docker_tag = format!("nousresearch/psyche-client:{}", client_version);
+        Ok(docker_tag)
+    }
+}
+
+/// Determine which Docker image to use and pull it if necessary
+async fn prepare_image(args: &Args, docker_mgr: &DockerManager) -> Result<String> {
+    let docker_tag = if let Some(ref local_image) = args.local {
+        info!("Using local Docker image: {}", local_image);
+        local_image.clone()
```

Comment:
for testing locally I think we should have the same logic here to check for the tag in the coordinator contract

Reply:
This should be done already, you can pass `--local` to fetch the coordinator version and use a local image, or `--local sometag` to force using `sometag` for testing.